### PR TITLE
Week2

### PR DESCRIPTION
**Pull request by ameliasheppy**
Hello! Finished my assignment for this week!

### CODE REVIEW

File: 01-node-tutorial/content/1-3-event_emitter.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
+console.log("Hello world!")
+const {readFile} = require('fs')
+const _ = require('lodash')
+
+const items = [1,[2,[3,[4]]]]
+const newItems = _.flattenDeep(items)
+
+console.log(newItems)
```
Comment (jdevries3133) [inline 01-node-tutorial/content/1-3-event_emitter.js:8]: I always like to see some experimentation outside the course content!

File: 01-node-tutorial/content/1-5-streams.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
+//streams are use to read and write big things:
+//writeable
+//readable
+//duplex
+//transform
```
Comment (jdevries3133) [inline 01-node-tutorial/content/1-5-streams.js:5]: What do you mean by these terms? Especially duplex and transform; I'm not sure I know what this means

File: 01-node-tutorial/content/errors-video.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+}
+console.log('...Then the execution continues')
+
+//you must use valid JS to get this block to work. Try catch only catches run time errors, not parse errors, so code must run to catch!
+
+//we can make custom errors messages.
+
+let json = '{"age":30}';
+try{
+    let user = JSON.parse(json);
```
Comment (jdevries3133) [inline 01-node-tutorial/content/errors-video.js:25]: I like that we're visiting the infamous JSON parse error. This is very common in the wild, if you keep your eyes peeled you can probably spot this happening in lots of websites

File: 01-node-tutorial/content/eventloopnotes.txt [hunk 4]
```diff
@@ -0,0 +1,32 @@
+The event loop is what let's Node do non-blocking ops
+despite the fact that JS is single-threaded!
+
+It offloads ops to the system kernel when possible. I'm fairly sure it's nothing popcorn-y.
```
Comment (jdevries3133) [inline 01-node-tutorial/content/eventloopnotes.txt:4]: Yes nothing popcorn-y unfortunately!

At a low level, the node runtime sends what are called "sys calls" to the operating system. Every operating system has a unique set of sys calls it supports, and node helps us write cross platform apps that mostly allow us to ignore that complexity!

Many syscalls are asynchronous, especially those dealing with files or the network, but this isn't node async / await; it's asynchrony at a much lower level, where the program is interacting with the operating system. 

This basically has the effect of allowing the program to continue doing it's thing while the OS uses other CPU cores on your machine to do the file or network I/O. Hence, by pushing this low level async work onto the operating system, node can stay busy evaluating JavaScript code. The JavaScript language async/await feature basically allows node to just evaluate a few lines until an OS call, then it switches to another task. That's why node is single-threaded but still async!

File: 01-node-tutorial/content/eventloopnotes.txt [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+We run a CL, then a setTimeout for 0. Then 100 CL. THe program will run:CL, 100 CL, then the setTimeout. 
+
+Let's build some code examples of it now!
+
+
+Don't forget, we have async and sync.
+Async is great for not blocking. But nested CB are the worst. So let's use a better solution!
+
+What code could we block the event loop with?
```
Comment (jdevries3133) [inline 01-node-tutorial/content/eventloopnotes.txt:28]: Can you answer this question with what you know now?

**Review by jdevries3133 — APPROVED** [review]
Nice work!

File: 01-node-tutorial/content/1-3-event_emitter.js (hunk 1 above)
Comment (ameliasheppy) [reply 01-node-tutorial/content/1-3-event_emitter.js:8]: Haha thanks! It is my best way to learn.

File: 01-node-tutorial/content/1-5-streams.js (hunk 2 above)
Comment (ameliasheppy) [reply 01-node-tutorial/content/1-5-streams.js:5]: In the video, John said that these are ways that we can do streams of data. Readable is used for read operations. (fs.createWriteStream())
Writable is used for write operations (fs.createReadStream()). 
Duplex are streams that are used for both read and write operations. 
An example that I found of the Duplex type is net.Socket. 
And transform is a type of duplex stream where the output is computed based on the input. 
I had made note of these to do further research on them. 
Also, with the transform type, I found an example that in the use of file compression, you can write data that is compressed and then read decompressed data to and from a file. zlib.createDeflate()

File: 01-node-tutorial/content/errors-video.js (hunk 3 above)
Comment (ameliasheppy) [reply 01-node-tutorial/content/errors-video.js:25]: Thanks for this tip!

File: 01-node-tutorial/content/eventloopnotes.txt (hunk 4 above)
Comment (ameliasheppy) [reply 01-node-tutorial/content/eventloopnotes.txt:4]: This is such a wonderful explanation!!

File: 01-node-tutorial/content/eventloopnotes.txt (hunk 5 above)
Comment (jdevries3133) [reply 01-node-tutorial/content/eventloopnotes.txt:28]: > An infinite loop? I'm pretty good at accidentally starting those, haha!

Yes, that's exactly the answer I was looking for!

File: 01-node-tutorial/content/1-5-streams.js (hunk 2 above)
Comment (jdevries3133) [reply 01-node-tutorial/content/1-5-streams.js:5]: 👍 very good

File: 01-node-tutorial/content/eventloopnotes.txt (hunk 4 above)
Comment (ameliasheppy) [reply 01-node-tutorial/content/eventloopnotes.txt:4]: Awesome info here! I have a background in medicine as a nurse, so computer science is all fairly new to me.  This response has given me a lot of research to do.

**Pull request merged**